src/optimistic_transaction_db.rs

```diff
@@ -0,0 +1,259 @@
+            Ok(c) => c,
+            Err(_) => {
+                return Err(Error::new(
+                    "Failed to convert path to CString \
+                     when opening DB."
+                        .to_owned(),
+                ));
+            }
+        };
+
+        if let Err(e) = fs::create_dir_all(&path) {
+            return Err(Error::new(format!(
+                "Failed to create RocksDB\
+                 directory: `{:?}`.",
+                e
+            )));
+        }
+
+        let db: *mut ffi::rocksdb_optimistictransactiondb_t;
+        let cf_map = Arc::new(RwLock::new(BTreeMap::new()));
+        let mut cfs: Vec<_> = cfs.into_iter().collect();
+        if cfs.is_empty() {
+            unsafe {
+                db = ffi_try!(ffi::rocksdb_optimistictransactiondb_open(
+                    opts.inner,
+                    cpath.as_ptr() as *const _,
+                ));
+            }
+        } else {
+            // Always open the default column family.
+            if !cfs
+                .iter()
+                .any(|cf: &ColumnFamilyDescriptor| cf.name == "default")
+            {
+                cfs.push(ColumnFamilyDescriptor {
+                    name: String::from("default"),
+                    options: Options::default(),
+                });
+            }
+            // We need to store our CStrings in an intermediate vector
+            // so that their pointers remain valid.
+            let c_cfs: Vec<CString> = cfs
+                .iter()
+                .map(|cf| CString::new(cf.name.as_bytes()).unwrap())
+                .collect();
+
+            let mut cf_names: Vec<_> = c_cfs.iter().map(|cf| cf.as_ptr()).collect();
+
+            // These handles will be populated by DB.
+            let mut cf_handles: Vec<_> = cfs.iter().map(|_| ptr::null_mut()).collect();
+
+            let mut cf_opts: Vec<_> = cfs.iter().map(|cf| cf.options.inner as *const _).collect();
+
+            unsafe {
+                db = ffi_try!(ffi::rocksdb_optimistictransactiondb_open_column_families(
+                    opts.inner,
+                    cpath.as_ptr(),
+                    cfs.len() as c_int,
+                    cf_names.as_mut_ptr(),
+                    cf_opts.as_mut_ptr(),
+                    cf_handles.as_mut_ptr(),
+                ));
+            }
+
+            for handle in &cf_handles {
+                if handle.is_null() {
+                    return Err(Error::new(
+                        "Received null column family \
+                         handle from DB."
+                            .to_owned(),
+                    ));
+                }
+            }
+
+            for (n, h) in cfs.iter().zip(cf_handles) {
+                cf_map
+                    .write()
+                    .map_err(|e| Error::new(e.to_string()))?
+                    .insert(n.name.clone(), h);
+            }
+        }
+
+        if db.is_null() {
```

Comment:
Won't `ffi_try!` on line 117 handle any situations where this pointer would be null? I ask because this is a very generic error message and wouldn't give any indication as to why the database could not be initialized.